celery/platforms.py

```diff
@@ -782,7 +782,7 @@ def check_privileges(accept_content):
                    for attr in ['getuid', 'getgid', 'geteuid', 'getegid']):
             raise SecurityError('suspicious platform, contact support')
 
-    if not uid or not gid or not euid or not egid:
+    if not uid or not euid:
```

Comment:
Per https://github.com/celery/celery/issues/6579#issuecomment-761858980, we may wish to check if the group's name is `wheel` or figure out which unix systems still rely on gid 0.
Celery supports a wide variety of deployment targets, even if not officially. People have production systems that we're unaware of.
We should make the effort to ensure no security issues occur from this bugfix.

Reply:
@thedrow how about we look at the group name as well as performing a UID/EUID check. Then if the group name is in a list of potential superuser privilege groups, we give the "`C_FORCE_ROOT`" related message and actions? Here's some sample code I've knocked up to demonstrate GID=1000 (a user defined group), GID=0 (root - which we'll allow) and GID=27 (sudo on my ubuntu instance).

```
#!/usr/bin/env python
# Python program to use grp.getgrgid() method to get group name from id. 
    
# import modules
import grp
import sys

# User and Effective User id
uid = 1000
euid = 1000

# Group id 
gid = 1000
egid = 1000

# Get the group database entry for the specified group id using grp.getgrgid() method
entry = grp.getgrgid(gid)

# Print the retrieved entry 
print("Group database entry for group id " + str(gid) + ": " + str(entry))

# Get and print group name based on gid
grp_entry = grp.getgrgid(gid)
grp_name = grp_entry[0]
print("Group Name for gid=" + str(gid) + ": " + grp_name)

if not uid or not euid:
    sys.stdout.write('User or Effective User ID is zero: uid=' + str(uid) + ' euid=' + str(euid) + ' gid=' + str(gid) + ' egid=' + str(egid) + '\n')
    sys.stdout.write('Running a worker with superuser privileges when the worker accepts messages serialized with pickle is a very bad idea!\n')
else:
    if str(grp_name) == 'sudo' or str(grp_name) == 'wheel':
        sys.stdout.write('No user IDs are zero: uid=' + str(uid) + ' euid=' + str(euid) + ' gid=' + str(gid) + ' egid=' + str(egid) + ', however group is: ' + grp_name + '\n')
        sys.stdout.write('Running a worker with superuser privileges when the worker accepts messages serialized with pickle is a very bad idea!\n')		
    else:
        sys.stdout.write('No user IDs are zero: uid=' + str(uid) + ' euid=' + str(euid) + ' gid=' + str(gid) + ' egid=' + str(egid) + '\n')
        sys.stdout.write('You are good to go!\n')

# =========================================================================================
# =========================================================================================
# =========================================================================================

# Group id 
gid = 0 # group is root
egid = 0
  
# Get the group database entry for the specified group id using grp.getgrgid() method 
entry = grp.getgrgid(gid) 

# Print the retrieved entry 
print("\nGroup database entry for group id " + str(gid) + ": " + str(entry))

# Get and print group name based on gid
grp_entry = grp.getgrgid(gid)
grp_name = grp_entry[0]
print("Group Name for gid=" + str(gid) + ": " + grp_name)

if not uid or not euid:
    sys.stdout.write('User or Effective User ID is zero: uid=' + str(uid) + ' euid=' + str(euid) + ' gid=' + str(gid) + ' egid=' + str(egid) + '\n')
    sys.stdout.write('Running a worker with superuser privileges when the worker accepts messages serialized with pickle is a very bad idea!\n')
else:
    if str(grp_name) == 'sudo' or str(grp_name) == 'wheel':
        sys.stdout.write('No user IDs are zero: uid=' + str(uid) + ' euid=' + str(euid) + ' gid=' + str(gid) + ' egid=' + str(egid) + ', however group is: ' + grp_name + '\n')
        sys.stdout.write('Running a worker with superuser privileges when the worker accepts messages serialized with pickle is a very bad idea!\n')	
    else:
        sys.stdout.write('No user IDs are zero: uid=' + str(uid) + ' euid=' + str(euid) + ' gid=' + str(gid) + ' egid=' + str(egid) + '\n')
        sys.stdout.write('You are good to go!\n')

# =========================================================================================
# =========================================================================================
# =========================================================================================

# Group id 
gid = 27 # group is sudo
egid = 27

# Get the group database entry for the specified group id using grp.getgrgid() method
entry = grp.getgrgid(gid) 

# Print the retrieved entry 
print("\nGroup database entry for group id " + str(gid) + ": " + str(entry))

# Get and print group name based on gid
grp_entry = grp.getgrgid(gid)
grp_name = grp_entry[0]
print("Group Name for gid=" + str(gid) + ": " + grp_name)

if not uid or not euid:
    sys.stdout.write('User or Effective User ID is zero: uid=' + str(uid) + ' euid=' + str(euid) + ' gid=' + str(gid) + ' egid=' + str(egid) + '\n')
    sys.stdout.write('Running a worker with superuser privileges when the worker accepts messages serialized with pickle is a very bad idea!\n')
else:
    if str(grp_name) == 'sudo' or str(grp_name) == 'wheel':
        sys.stdout.write('No user IDs are zero: uid=' + str(uid) + ' euid=' + str(euid) + ' gid=' + str(gid) + ' egid=' + str(egid) + ', however group is: ' + grp_name + '\n')
        sys.stdout.write('Running a worker with superuser privileges when the worker accepts messages serialized with pickle is a very bad idea!\n')	
    else:
        sys.stdout.write('No user IDs are zero: uid=' + str(uid) + ' euid=' + str(euid) + ' gid=' + str(gid) + ' egid=' + str(egid) + '\n')
        sys.stdout.write('You are good to go!\n')

# =========================================================================================
# =========================================================================================
# =========================================================================================
```

The output of this script on my Ubuntu machine is:
```
Group database entry for group id 1000: grp.struct_group(gr_name='matt', gr_passwd='x', gr_gid=1000, gr_mem=[])
Group Name for gid=1000: matt
No user IDs are zero: uid=1000 euid=1000 gid=1000 egid=1000
You are good to go!

Group database entry for group id 0: grp.struct_group(gr_name='root', gr_passwd='x', gr_gid=0, gr_mem=[])
Group Name for gid=0: root
No user IDs are zero: uid=1000 euid=1000 gid=0 egid=0
You are good to go!

Group database entry for group id 27: grp.struct_group(gr_name='sudo', gr_passwd='x', gr_gid=27, gr_mem=['matt'])
Group Name for gid=27: sudo
No user IDs are zero: uid=1000 euid=1000 gid=27 egid=27, however group is: sudo
Running a worker with superuser privileges when the worker accepts messages serialized with pickle is a very bad idea!
```

And I think this then gives the ability to define a list of groups by name (rather than ID which can be different based on distribution), that we are or aren't happy with as default and so require `C_FORCE_ROOT` override.

(The code might need some refactoring to also check `egid` against the list of named groups as well, but in principle it'll follow the same check as against `gid`).

Thoughts?

Reply:
That should do the trick.
Note that these functions may raise a `KeyError` and in that case we should failover to mandating `C_FORCE_ROOT` with an appropriate log message.

I'm being particularly meticulous here to avoid potential security issues.

Reply:
Yep, can completely understand and appreciate that.

Reply:
@thedrow I've updated the code. Can you please explain where you feel the `KeyError` may come in and specify the code you want to handle it so I can include it?

Reply:
He probably metioned def performTest(uid,euid,gid,egid):
    # Get the group database entry for the specified group id using grp.getgrgid() method
    try:
        entry = grp.getgrgid(gid)
    except KeyError:
        print('gid or egid not found')
        return  # 'exit' function and return to caller

and suggested to fallback to or failover to mandating C_FORCE_ROOT with an appropriate log message.

Reply:
@thedrow and @auvipy I've added in the `KeyError` handler please can you review.